File: ios/TikTokBusinessModule.swift
import Foundation
import React
import UIKit
import TikTokBusinessSDK

@objc(TikTokBusinessModule)
class TikTokBusinessModule: NSObject, RCTBridgeModule {
  
  // MARK: - RCTBridgeModule Protocol
  
  static func moduleName() -> String! {
    return "TikTokBusinessModule"
  }
  
  static func requiresMainQueueSetup() -> Bool {
    return false
  }
  
  // MARK: - Public Methods Exposed to JS
  
  /// Identifies the user (login, signup, profile update, etc.).
  @objc func identify(_ externalId: String,
                      externalUserName: String?,
                      phoneNumber: String?,
                      email: String,
                      resolver: @escaping RCTPromiseResolveBlock,
                      rejecter: @escaping RCTPromiseRejectBlock) {
    do {
      TikTokBusiness.identify(withExternalID: externalId,
                              externalUserName: externalUserName,
                              phoneNumber: phoneNumber,
                              email: email)
      resolver("User identified successfully")
    } catch {
      rejecter("IDENTIFY_ERROR", "Failed to identify user", error)
    }
  }
  
  /// Logs out the user.
  @objc func logout(_ resolver: @escaping RCTPromiseResolveBlock,
                    rejecter: @escaping RCTPromiseRejectBlock) {
    do {
      TikTokBusiness.logout()
      resolver("User logged out successfully")
    } catch {
      rejecter("LOGOUT_ERROR", "Failed to logout user", error)
    }
  }
  
  /// Reports a standard event.
  /// If `parameters` is nil or empty, the event is reported without additional properties.
  /// Otherwise, a TTBaseEvent is built with the provided properties.
  /// Expects the eventName to match one of the values in the EventName enum.
  @objc func trackEvent(_ eventName: String,
                        eventId: String?,
                        parameters: NSDictionary?,
                        resolver: @escaping RCTPromiseResolveBlock,
                        rejecter: @escaping RCTPromiseRejectBlock) {
    do {
      let event = TikTokBaseEvent(eventName: eventName, eventId: eventId)
      
      if parameters != nil && (parameters?.count ?? 0) > 0 {
        for (key, value) in parameters! {
          guard let keyStr = key as? String else { continue }
          if let stringValue = value as? String {
            event.addProperty(withKey: keyStr, value: stringValue)
          } else if let numberValue = value as? NSNumber {
            event.addProperty(withKey: keyStr, value: Double(numberValue))
          } else if let boolValue = value as? Bool {
            event.addProperty(withKey: keyStr, value: boolValue)
          }
        }
      }
      
      TikTokBusiness.trackTTEvent(event)
      resolver("Event tracked successfully")
    } catch {
      rejecter("TRACK_EVENT_ERROR", "Failed to track event", error)
    }
  }
  
  /// Reports a content event.
  /// Selects the appropriate event builder based on the eventType and iterates over the "CONTENTS" array to add all content items.
  @objc func trackContentEvent(_ eventType: String,
                               properties: NSDictionary?,
                               resolver: @escaping RCTPromiseResolveBlock,
                               rejecter: @escaping RCTPromiseRejectBlock) {
    do {
      let event: TikTokContentsEvent
      switch eventType {
      case "ADD_TO_CART":
        event = TikTokAddToCartEvent()
      case "ADD_TO_WISHLIST":
        event = TikTokAddToWishlistEvent()
      case "CHECK_OUT":
        event = TikTokCheckoutEvent()      
      case "PURCHASE":
        event = TikTokPurchaseEvent()
      case "VIEW_CONTENT":
        event = TikTokViewContentEvent()
      default:
        rejecter("UNSUPPORTED_EVENT_TYPE", "Unsupported content event type: \(eventType)", nil)
        return
      }
      
      if let props = properties, props.count > 0 {
        if let description = props["DESCRIPTION"] as? String {
          event.setDescription(description)
        }
        if let currencyStr = props["CURRENCY"] as? String {
          // Map currency string to TTCurrency enum - using USD as default
          let currency = TTCurrency.USD // You may want to add proper currency mapping
          event.setCurrency(currency)
        }
        if let value = props["VALUE"] as? NSNumber {
          event.setValue(value.stringValue)
        }
        if let contentType = props["CONTENT_TYPE"] as? String {
          event.setContentType(contentType)
        }
        if let contentId = props["CONTENT_ID"] as? String {
          event.setContentId(contentId)
        }
        
        // Process the "CONTENTS" array
        if let contentsArray = props["CONTENTS"] as? [NSDictionary] {
          var contentParamsList : [TikTokContentParams] = []
          for contentDict in contentsArray {
            let eventContent = TikTokContentParams()
            if let contentId = contentDict["CONTENT_ID"] as? String {
              eventContent.contentId = contentId
            }
            if let brand = contentDict["BRAND"] as? String {
              eventContent.brand = brand
            }
            if let price = contentDict["PRICE"] as? NSNumber {
              eventContent.price = NSNumber(value: price.doubleValue)
            }
            if let quantity = contentDict["QUANTITY"] as? NSNumber {
              eventContent.quantity = quantity.intValue
            }
            if let contentName = contentDict["CONTENT_NAME"] as? String {
              eventContent.contentName = contentName
            }
            contentParamsList.append(eventContent)
          }
          // Set all content items using the varargs setter
          event.setContents(contentParamsList)
        }
      }
      
      TikTokBusiness.trackTTEvent(event)
      resolver("Content event tracked successfully")
    } catch {
      rejecter("TRACK_CONTENT_EVENT_ERROR", "Failed to track content event", error)
    }
  }
  
  /// Reports a custom event.
  /// Builds the event using TTBaseEvent.newBuilder and adds the provided properties.
  @objc func trackCustomEvent(_ eventName: String,
                              properties: NSDictionary?,
                              resolver: @escaping RCTPromiseResolveBlock,
                              rejecter: @escaping RCTPromiseRejectBlock) {
    do {
      let customEvent = TikTokBaseEvent(name: eventName)
      if let props = properties, props.count > 0 {
        for (key, value) in props {
          guard let keyStr = key as? String else { continue }
          if let stringValue = value as? String {
            customEvent.addProperty(withKey: keyStr, value: stringValue)
          } else if let numberValue = value as? NSNumber {
            customEvent.addProperty(withKey: keyStr, value: numberValue)
          } else if let boolValue = value as? Bool {
            customEvent.addProperty(withKey: keyStr, value: boolValue)
          }
        }
      }
      
      TikTokBusiness.trackTTEvent(customEvent)
      resolver("Custom event tracked successfully")
    } catch {
      rejecter("TRACK_CUSTOM_EVENT_ERROR", "Failed to track custom event", error)
    }
  }
  
  /// Reports an ad revenue event.
  /// Uses TikTokAdRevenueEvent to track impression-level ad revenue data.
  @objc func trackAdRevenueEvent(_ adRevenueJson: NSDictionary,
                                 eventId: String?,
                                 resolver: @escaping RCTPromiseResolveBlock,
                                 rejecter: @escaping RCTPromiseRejectBlock) {
    do {
      guard let adRevenueDict = adRevenueJson as? [AnyHashable: Any] else {
        rejecter("INVALID_AD_REVENUE_DATA", "Failed to convert ad revenue data", nil)
        return
      }
      
      let adRevenueEvent: TikTokAdRevenueEvent
      if let eventId = eventId, !eventId.isEmpty {
        adRevenueEvent = TikTokAdRevenueEvent(adRevenue: adRevenueDict, eventId: eventId)
      } else {
        adRevenueEvent = TikTokAdRevenueEvent(adRevenue: adRevenueDict, eventId: "")
      }
      
      TikTokBusiness.trackTTEvent(adRevenueEvent)
      resolver("Ad revenue event tracked successfully")
    } catch {
      rejecter("TRACK_AD_REVENUE_ERROR", "Failed to track ad revenue event", error)
    }
  }
  
  /// Initializes the TikTok SDK.
  /// Accepts appId, ttAppId, accessToken, and an optional debug flag.
  @objc func initializeSdk(_ appId: String, 
                          ttAppId: String, 
                          accessToken: String, 
                          debug: NSNumber, 
                          resolver: @escaping RCTPromiseResolveBlock,
                          rejecter: @escaping RCTPromiseRejectBlock) {
    let config = TikTokConfig(accessToken: accessToken, appId: appId, tiktokAppId: ttAppId)
    
    let debugValue = debug.boolValue
    if debugValue {
      config?.enableDebugMode()
      config?.setLogLevel(TikTokLogLevelDebug)
    }
    
    TikTokBusiness.initializeSdk(config) { success, error in
      if success {
        print("[TikTokBusiness] TikTokBusiness initialized OK")
        resolver("SDK initialized successfully")
      } else {
        print("[TikTokBusiness] Initialization failed: \(error?.localizedDescription ?? "Unknown error")")
        rejecter("INIT_ERROR", "Failed to initialize TikTok SDK", error)
      }
    }
  }

  @objc func flush() {
    TikTokBusinessSdk.explicitlyFlush()
  }
}

File: android/src/main/java/com/tiktokbusiness/TikTokBusinessModule.kt
package com.tiktokbusiness

import com.facebook.react.bridge.*
import com.tiktok.TikTokBusinessSdk
import com.tiktok.appevents.base.EventName
import com.tiktok.appevents.base.TTAdRevenueEvent
import com.tiktok.appevents.base.TTBaseEvent
import com.tiktok.appevents.contents.*
import org.json.JSONObject

class TikTokBusinessModule(reactContext: ReactApplicationContext) :
  ReactContextBaseJavaModule(reactContext) {

  override fun getName(): String {
    return NAME
  }

  /**
   * Identifies the user when user information changes (login, signup, profile update, etc.).
   */
  @ReactMethod
  fun identify(
    externalId: String, externalUserName: String?, phoneNumber: String?, email: String?,
    promise: Promise
  ) {
    try {
      TikTokBusinessSdk.identify(externalId, externalUserName, phoneNumber, email)
      promise.resolve("User identified successfully")
    } catch (e: Exception) {
      promise.reject("IDENTIFY_ERROR", "Failed to identify user", e)
    }
  }

  /**
   * Logs out the user.
   */
  @ReactMethod
  fun logout(promise: Promise) {
    try {
      TikTokBusinessSdk.logout()
      promise.resolve("User logged out successfully")
    } catch (e: Exception) {
      promise.reject("LOGOUT_ERROR", "Failed to logout user", e)
    }
  }

  /**
   * Maps event name strings to EventName enums.
   */
  private fun getEventNameEnum(eventName: String): EventName? {
    return EventName.values().find { it.toString() == eventName }
  }

  /**
   * Reports a standard event.
   * If 'parameters' is null or empty, the event is reported without additional properties.
   * Otherwise, it builds a TTBaseEvent with the provided properties.
   *
   * Expects the eventName to match one of the string values in the EventName enum.
   */
  @ReactMethod
  fun trackEvent(
    eventName: String, eventId: String?, parameters: ReadableMap?, promise: Promise
  ) {
    try {
      val eventEnum = getEventNameEnum(eventName)
      if (eventEnum == null) {
        promise.reject("INVALID_EVENT_NAME", "Unknown event name: $eventName", null)
        return
      }
      if (parameters == null || !parameters.keySetIterator().hasNextKey()) {
        if (eventId.isNullOrEmpty()) {
          TikTokBusinessSdk.trackTTEvent(eventEnum)
        } else {
          TikTokBusinessSdk.trackTTEvent(eventEnum, eventId)
        }
      } else {
        val builder = TTBaseEvent.newBuilder(eventEnum.toString())
        val iterator = parameters.keySetIterator()
        while (iterator.hasNextKey()) {
          val key = iterator.nextKey()
          when (parameters.getType(key)) {
            ReadableType.String -> builder.addProperty(key, parameters.getString(key))
            ReadableType.Number -> builder.addProperty(key, parameters.getDouble(key))
            ReadableType.Boolean -> builder.addProperty(key, parameters.getBoolean(key))
            else -> {
              // Ignore other types if not needed
            }
          }
        }
        val event = builder.build()
        TikTokBusinessSdk.trackTTEvent(event)
      }
      promise.resolve("Event tracked successfully")
    } catch (e: Exception) {
      promise.reject("TRACK_EVENT_ERROR", "Failed to track event", e)
    }
  }

  /**
   * Reports a content event.
   * Selects the appropriate event builder based on the eventType and builds the event
   * using the provided properties. Iterates through the "CONTENTS" array and aggregates all items.
   */
  @ReactMethod
  fun trackContentEvent(eventType: String, properties: ReadableMap?, promise: Promise) {
    try {
      val builder: TTContentsEvent.Builder = when (eventType) {
        "ADD_TO_CART" -> TTAddToCartEvent.newBuilder()
        "ADD_TO_WISHLIST" -> TTAddToWishlistEvent.newBuilder()
        "CHECK_OUT" -> TTCheckoutEvent.newBuilder()
        "PURCHASE" -> TTPurchaseEvent.newBuilder()
        "VIEW_CONTENT" -> TTViewContentEvent.newBuilder()
        else -> {
          promise.reject("UNSUPPORTED_EVENT_TYPE", "Unsupported content event type: $eventType", null)
          return
        }
      }
      if (properties != null && properties.keySetIterator().hasNextKey()) {
        if (properties.hasKey("DESCRIPTION") && properties.getType("DESCRIPTION") == ReadableType.String) {
          builder.setDescription(properties.getString("DESCRIPTION"))
        }
        if (properties.hasKey("CURRENCY") && properties.getType("CURRENCY") == ReadableType.String) {
          val currencyStr = properties.getString("CURRENCY")
          if (!currencyStr.isNullOrEmpty()) {
            builder.setCurrency(TTContentsEventConstants.Currency.valueOf(currencyStr))
          }
        }
        if (properties.hasKey("VALUE") && properties.getType("VALUE") == ReadableType.Number) {
          builder.setValue(properties.getDouble("VALUE"))
        }
        if (properties.hasKey("CONTENT_TYPE") && properties.getType("CONTENT_TYPE") == ReadableType.String) {
          builder.setContentType(properties.getString("CONTENT_TYPE"))
        }
        // Iterate over the CONTENTS array and accumulate all content items.
        if (properties.hasKey("CONTENTS") && properties.getType("CONTENTS") != ReadableType.Null) {
          val contentsArray = properties.getArray("CONTENTS")
          if (contentsArray != null && contentsArray.size() > 0) {
            val contentsList = mutableListOf<TTContentParams>()
            for (i in 0 until contentsArray.size()) {
              val contentMap = contentsArray.getMap(i)
              if (contentMap != null) {
                val contentBuilder = TTContentParams.newBuilder()
                if (contentMap.hasKey("CONTENT_ID") && contentMap.getType("CONTENT_ID") == ReadableType.String) {
                  contentBuilder.setContentId(contentMap.getString("CONTENT_ID"))
                }
                if (contentMap.hasKey("CONTENT_CATEGORY") && contentMap.getType("CONTENT_CATEGORY") == ReadableType.String) {
                  contentBuilder.setContentCategory(contentMap.getString("CONTENT_CATEGORY"))
                }
                if (contentMap.hasKey("BRAND") && contentMap.getType("BRAND") == ReadableType.String) {
                  contentBuilder.setBrand(contentMap.getString("BRAND"))
                }
                if (contentMap.hasKey("PRICE") && contentMap.getType("PRICE") == ReadableType.Number) {
                  contentBuilder.setPrice(contentMap.getDouble("PRICE").toFloat())
                }
                if (contentMap.hasKey("QUANTITY") && contentMap.getType("QUANTITY") == ReadableType.Number) {
                  contentBuilder.setQuantity(contentMap.getInt("QUANTITY"))
                }
                if (contentMap.hasKey("CONTENT_NAME") && contentMap.getType("CONTENT_NAME") == ReadableType.String) {
                  contentBuilder.setContentName(contentMap.getString("CONTENT_NAME"))
                }
                contentsList.add(contentBuilder.build())
              }
            }
            // Set all content items via varargs
            builder.setContents(*contentsList.toTypedArray())
          }
        }
      }
      val event = builder.build()
      TikTokBusinessSdk.trackTTEvent(event)
      promise.resolve("Content event tracked successfully")
    } catch (e: Exception) {
      promise.reject("TRACK_CONTENT_EVENT_ERROR", "Failed to track content event", e)
    }
  }

  /**
   * Reports a custom event.
   * Builds the event using TTBaseEvent.newBuilder and adds the provided properties.
   */
  @ReactMethod
  fun trackCustomEvent(eventName: String, properties: ReadableMap?, promise: Promise) {
    try {
      val builder = TTBaseEvent.newBuilder(eventName)
      if (properties != null && properties.keySetIterator().hasNextKey()) {
        val iterator = properties.keySetIterator()
        while (iterator.hasNextKey()) {
          val key = iterator.nextKey()
          when (properties.getType(key)) {
            ReadableType.String -> builder.addProperty(key, properties.getString(key))
            ReadableType.Number -> builder.addProperty(key, properties.getDouble(key))
            ReadableType.Boolean -> builder.addProperty(key, properties.getBoolean(key))
            else -> {
              // Ignore other types if not needed
            }
          }
        }
      }
      val event = builder.build()
      TikTokBusinessSdk.trackTTEvent(event)
      promise.resolve("Custom event tracked successfully")
    } catch (e: Exception) {
      promise.reject("TRACK_CUSTOM_EVENT_ERROR", "Failed to track custom event", e)
    }
  }

  /**
   * Reports an ad revenue event.
   * Uses TTAdRevenueEvent to track impression-level ad revenue data.
   */
  @ReactMethod
  fun trackAdRevenueEvent(adRevenueJson: ReadableMap, eventId: String?, promise: Promise) {
    try {
      val jsonObject = JSONObject()
      val iterator = adRevenueJson.keySetIterator()
      while (iterator.hasNextKey()) {
        val key = iterator.nextKey()
        when (adRevenueJson.getType(key)) {
          ReadableType.String -> jsonObject.put(key, adRevenueJson.getString(key))
          ReadableType.Number -> jsonObject.put(key, adRevenueJson.getDouble(key))
          ReadableType.Boolean -> jsonObject.put(key, adRevenueJson.getBoolean(key))
          else -> {
            // Skip unsupported types
          }
        }
      }

      val event = if (eventId.isNullOrEmpty()) {
        TTAdRevenueEvent.newBuilder(jsonObject).build()
      } else {
        TTAdRevenueEvent.newBuilder(jsonObject, eventId).build()
      }

      TikTokBusinessSdk.trackTTEvent(event)
      promise.resolve("Ad revenue event tracked successfully")
    } catch (e: Exception) {
      promise.reject("TRACK_AD_REVENUE_ERROR", "Failed to track ad revenue event", e)
    }
  }

  @ReactMethod
  fun initializeSdk(appId: String, ttAppId: String, accessToken: String, debug: Boolean?, promise: Promise) {
    try {
      val configBuilder =
        TikTokBusinessSdk.TTConfig(reactApplicationContext, accessToken).setAppId(appId).setTTAppId(ttAppId)

      if (debug == true) {
        configBuilder.openDebugMode().setLogLevel(TikTokBusinessSdk.LogLevel.DEBUG)
      }

      TikTokBusinessSdk.initializeSdk(configBuilder, object : TikTokBusinessSdk.TTInitCallback {
        override fun success() {
          TikTokBusinessSdk.startTrack()
          promise.resolve("SDK initialized successfully")
        }

        override fun fail(code: Int, msg: String) {
          promise.reject("INIT_ERROR", "Failed to initialize TikTok SDK: $msg (code: $code)", null)
        }
      })
    } catch (e: Exception) {
      promise.reject("INIT_ERROR", "Failed to initialize TikTok SDK", e)
    }
  }

  companion object {
    const val NAME = "TikTokBusinessModule"
  }
}
